cachi2/interface/cli.py

```diff
@@ -22,6 +22,7 @@
 log = logging.getLogger(__name__)
 
 DEFAULT_SOURCE = "."
+DEFAULT_SOURCE_BACKUP = "./cachi2-source"
```

Comment:
Static paths created and owned by a regular user (which cachi2 runs as) are predictable and thus error prone (e.g. DOS attack), so we always need to default to a directory with a random name.

Moreover, while it is perfectly fine to default to `./`  for the reflink use case because only metadata is copied and not the actual blocks, we should default to a directory residing on a pseudo filesystem on the fallback path to reduce wear on the hard drive since we'd be writing lots of small files with every project request.

Reply:
I think it's fine to use a random name for the dir, but note that we already use a static path for the default output dir currently.

On the second matter, I think my knowledge is lacking here. What do you mean by pseudo filesystem, and how would we go about making this copy in one?

Reply:
> I think it's fine to use a random name for the dir, but note that we already use a static path for the default output dir currently.

That's different though, because outputdir is user facing, the copied dir proppsed in this PR is not. However, we might want to look at writing the output dir atomically.

> What do you mean by pseudo filesystem
pseudo filesystem == not a physical file system, it's purely RAM-backed, e.g. tmpfs, devfs, procfs, etc.

Reply:
My initial approach was to copy it to `/tmp`, but I bumped into a `ReflinkImpossibleError: EXDEV` error since it identifies it as cross device. This is one of the reasons I thought it would be wise to allow the user to force-disable reflinks.  I'm on `6.4.10-200.fc38.x86_64`, btw.

Reply:
Yeah, that is expected, so that's why I suggested to go with a random name under `./` with reflinks as that should go smooth as butter™ and only copy to something like `/tmp` (it's not a tmpfs everywhere, but I guess very few projects actually check if it is to reduce the wear anyway) on fallback.

Reply:
I think the question about where to write is very environment-dependent, which (in my mind) leads to the question of what environment(s) we want to fully support. Local (bare metal) mode is fine for development, but I'm not sure that I see a *useful* use case otherwise.

All of which leads me to

> reduce wear on the hard drive since we'd be writing lots of small files with every project request

which *really* feels like
1. premature optimization
2. not our problem, given that our *primary* "production" use case is in virtualized or containerized environments

Reply:
I kept the copy location inside the current `source_dir` for simplicity reasons. If we figure out a way to make use of `/tmp`, I can update this PR, otherwise we can leave it as an improvement later.

Reply:
What is worrying me now with the copy approach is that we might have size limitations in the `workspace` volume in RHTAP, which could essentially block users that want to build larger Yarn repos. I'll try to figure out more about size constraints tomorrow.

Reply:
> which really feels like
>
>   1. premature optimization
>   2. not our problem, given that our primary "production" use case is in virtualized or containerized environments

I don't think it is premature at all, the problem is very real. While the general mindset in production clouds is that the industry doesn't care about storage wear - true, but local deployments do care very much. Again, we need to differentiate what a given company's expectations wrt/ this project are vs what is important/beneficial for upstream community in general and how these two can align. I understand that from production POV we'll probably only ever see containerized environments, but development environments should not be in a disadvantage. IOW I as a developer who prefers local deployments for instant feedback loop (let's face it CI/CD is designed to get significantly slower with time when it comes to feedback loop) very much care about not wearing down my SSD prematurely, as it's my day-in-day-out job to test my code (and others' too). So, I personally am not convinced about downplaying the development use case by saying it's not "fully" supported.

That said, I won't stand in the way of always copying to `./` and let end users deal with the fact they need to pick different file system for their repo (like it would be my case going from EXT4 to e.g. BTRFS just for git repos) if the overall consensus is that we don't want to special case this.

Another alternative I could think of  would be re-visiting what @brunoapimentel proposed originally with a CLI option, but instead calling it `--backup-dir` and pitching this as a directory where we copy the whole project and then remove it before the user could even touch it, we could have a more generic CLI option `--tmpdir` and pitch this as an "alternative directory for cachi2 to use for its temporary data which depending on the project could be very large, surpassing the capabilities of /tmp" instead. We're not giving away any internal implementation detail, it's as vague as possible and nobody should expect any kind of stability in that regard.

Reply:
> What is worrying me now with the copy approach is that we might have size limitations in the workspace volume in RHTAP, which could essentially block users that want to build larger Yarn repos. I'll try to figure out more about size constraints tomorrow.

This constraint exists outside of cachi2 though, so not something we should concern ourselves with in this project and in this PR IMO, but yes, still worth investigating what size constraints we're dealing with in that specific deployment.

Reply:
In hindsight I think we can add a global (i.e. package manager agnostic) CLI option `--tmpdir` in a follow up PR and not block this one. However, I still think we should keep in mind local environments and frequent integration tests where we shouldn't copy/write unnecessary data.